Update Popup explainer to incorporate all recent changes (#527)
openui/open-ui: research/src/pages/popup/popup.research.explainer.mdx

  - [Disallowed elements](#disallowed-elements)
- [Example Use Cases](#example-use-cases)
  - [Generic Popup (Date Picker)](#generic-popup-date-picker)
  - [Generic popup (`<selectmenu>` listbox example)](#generic-popup-selectmenu-listbox-example)
  - [Hint/Tooltip](#hinttooltip)
  - [Async](#async)
- [Additional Considerations](#additional-considerations)
  - [Exceeding the Frame Bounds](#exceeding-the-frame-bounds)
  - [Shadow DOM](#shadow-dom)
  - [Eventual Single-Purpose Elements](#eventual-single-purpose-elements)
- [The Choices Made in this API](#the-choices-made-in-this-api)
  - [Alternatives Considered](#alternatives-considered)
  - [Design decisions](#design-decisions)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Background

A very common UI pattern on the Web, for which there is no native API, is "popup UI" or "popups". Popups are a general class of UI that have three common behaviors:
 1. Popups always appear **on top of other page content**.
 2. Popups are **ephemeral**. When the user "moves on" to another part of the page (e.g. by clicking elsewhere, or hitting ESC), the popup closes.

>  3. Popups are **"one at a time"** - opening one popup closes others.

Author: Yep, done, thanks!

### Page Load Trigger

As mentioned above, a `<div popup=popup>` will be hidden by default. If it is desired that the popup should be shown automatically upon page load, the `defaultopen` attribute can be applied:

```html
<div popup=popup defaultopen>
```

In this case, the UA will immediately call `showPopup()` on the element, as it is parsed. If multiple such elements exist on the page, only the first such element (in DOM order) on the page will be shown.

Note that `hint` popups cannot use the `defaultopen` attribute:

```html
<div popup=hint defaultopen>I will not be shown on page load</div>
```

Note also that more than one `async` popup can use `defaultopen` and all such popups will be shown on load, not just the first one:

```html
<div popup=async defaultopen>Shown on page load</div>
<div popup=async defaultopen>Also shown on page load</div>
```

### Shown vs Hidden Popups

The styling for a popup is provided by roughly the following UA stylesheet rules:

```css
[popup i]:not(:popup-open) {
  display: none;
}

[popup i] {
  position: fixed;
}
```

> The above rules mean that a popup, when not "shown", has `display:none` applied, and that style is removed when one of the methods above is used to show the popup. Note that the `display:none` UA stylesheet rule is **not** `!important`. In other words, developer style rules can be used to override this UA style to make a not-showing popup visibile in the page. In this case, the popup will **not** be displayed in the top layer, but rather at it's ordinary z-index position within the document. This can be used, for example, to animate the show/hide behavior of the popup, or make popup content "return to the page" instead of becoming hidden.

Reviewer: ```suggestion
The above rules mean that a popup, when not "shown", has `display:none` applied, and that style is removed when one of the methods above is used to show the popup. Note that the `display:none` UA stylesheet rule is **not** `!important`. In other words, developer style rules can be used to override this UA style to make a not-showing popup visible in the page. In this case, the popup will **not** be displayed in the top layer, but rather at it's ordinary `z-index` position within the document. This can be used, for example, to animate the show/hide behavior of the popup, or make popup content "return to the page" instead of becoming hidden.
```

Author: Thanks! done.

}
```

Further, only [valid values](#html-content-attribute) of the content attribute will be reflected to the IDL property, with invalid values being reflected as the empty string `""`. For example:

```javascript
const div = document.createElement('div');
div.setAttribute('popup','hint');
div.popup === 'hint'; // true
div.setAttribute('popup','invalid!');
div.popup === ''; // true
```

## Events

Events are fired (asynchronously) when a popup is shown (`show` event) and hidden (`hide` event). These events can be used, for example, to populate content for the popup just in time before it is shown, or update server data when it closes. The events are:

```javascript
const popup = Object.assign(document.createElement('div'), {popup: 'popup'});
popup.addEventListener('show',() => console.log('Popup is being shown!'));
popup.addEventListener('hide',() => console.log('Popup is being hidden!'));
```

Neither of these events are cancellable, and both are fired asynchronously.

## Focus Management

> Elements that move into the top layer sometimes need to move the focus to that element, and sometimes don't. For example, a modal `<dialog>` gets automatically focused because a dialog is something that requires immediate focus and attention. On the other hand, a `<div popup=hint>` doesn't receive focus at all (and is typically not expected to contain focusable elements). Similarly, a `<div popup=async>` should not receive focus (even if focusable) because it is meant for out-of-band communication of state, and should not interrupt a user's current action. Additionally, if the top layer element **should** receive immediate focus, there is a question about **which** part of the element gets that initial focus. For example, the element itself could receive focus, or one of its focusable descendants could receive focus first. To provide control over these behaviors, two attributes can be used on popups:

Author: I like it. Not sure why the diff is terrible, but I got it.

> - `autofocus`. When present on a popup or the descendant of a popup, this causes focus to be moved to that element when the popup is shown. Note that `autofocus` is [already a global attribute](https://html.spec.whatwg.org/multipage/interaction.html#the-autofocus-attribute), but the existing behavior applies to element focus on **page load**. This proposal extends that definition to be used within popups, and the focus behavior happens **when they are shown**. Note that adding `autofocus` to a popup descendant does **not** cause the popup to be shown on page load, and therefore it does not cause focus to be moved into the popup **on page load**, unless the `defaultopen` attribute is also used.

Reviewer: note that this use of `autofocus` is also relevant to the initial focus placement doc/wiki that was worked on for the dialog element.  so there is pending precedent for this proposal for the attribute.

```suggestion
- `autofocus`. When present on a popup or one of its descendants, it will result in focus being moved to the specified element when the popup is rendered. Note that `autofocus` is [already a global attribute](https://html.spec.whatwg.org/multipage/interaction.html#the-autofocus-attribute), but the existing behavior applies to element focus on **page load**. This proposal extends that definition to be used within popups, and the focus behavior happens **when they are shown**. Note that adding `autofocus` to a popup descendant does **not** cause the popup to be shown on page load, and therefore it does not cause focus to be moved into the popup **on page load**, unless the `defaultopen` attribute is also used.
```

Author: Accepted.

> - `delegatesfocus`. When present on the popup element itself, this causes the **first focusable descendent** of the popup to be focused when the popup is shown. (Eventually, this attribute might be made applicable to any element, not just popups, to control focus behavior more generally.)

Reviewer: Would it be good to call out that the purpose of such an attribute would be for instances where the subtree of the popup may not be known prior to its rendering?  Otherwise, it seems like something that specifying `autofocus` on the first focusable element of a popup could have taken care of?

Author: Yep, good point. Done.

```html
<div popup=popup>I'm a popup</div>
<style>
[popup]::backdrop {
  backdrop-filter: blur(3px);
}
</style>
```

Note that in contrast to the `::backdrop` pseudo element for modal dialogs and fullscreen elements, the `::backdrop` for a popup is styled by a UA stylesheet rule `pointer-events: none !important`, which means it cannot trap clicks outside the popup. This ensures the "click outside" light dismiss behavior continues to function.

# Behaviors

## Automatic Dismiss Behavior

For elements that are displayed on the top layer via this API, there are a number of things that can cause the element to be removed from the top-layer, besides the ones [described above](#showing-and-hiding-a-popup). These fall into three main categories:

* **One at a Time.** Another element being added to the top-layer causes existing top layer elements to be removed. This is typically used for “one at a time” type elements: when one popup is shown, other popups should be hidden, so that only one is on-screen at a time. This is also used when “more important” top layer elements get added. For example, fullscreen elements should close all open popups.

> * **Light Dismiss**. User action such as clicking outside the element, hitting Escape, or causing keyboard focus to leave the element should all cause a showing popup to be hidden. This is typically called “light dismiss”, and is discussed in more detail in [this section](#light-dismiss).

Reviewer: just a consistency nit, putting the period inside the bold like the previous

```suggestion
* **Light Dismiss.** User action such as clicking outside the element, hitting Escape, or causing keyboard focus to leave the element should all cause a displayed popup to be hidden. This is typically called “light dismiss”, and is discussed in more detail in [this section](#light-dismiss).
```

Author: Thanks! I hate when those aren't consistent.

On the `displayed` vs `showing`, I fought with that throughout. I wanted to avoid "open"/"closed" because we're using "show"/"hide", but it's awkward in many cases. I like "displayed", I might try to use that more.

> * **Other Reasons**. Because the top layer is a UA-managed resource, it may have other reasons (for example a user preference) to forcibly remove elements from the top layer.

Reviewer: same same
```suggestion
* **Other Reasons.** Because the top layer is a UA-managed resource, it may have other reasons (for example a user preference) to forcibly remove elements from the top layer.
```

In all such cases, the UA is allowed to forcibly remove an element from the top layer and re-apply the `display:none` popup UA rule. The rules the UA uses to manage these interactions depends on the element types, and this is described in more detail in [this section](#classes-of-top-layer-ui).

### Light Dismiss

The term "light dismiss" for a popup is used to describe the user "moving on" to another part of the page, or generally being done interacting with the popup. When this happens, the popup should be hidden. Several actions trigger light dismiss of a popup:

> 1. Clicking or tapping outside the bounds of the popup. Any `mousedown` event will trigger **all open popups** to be hidden, starting from the top of [the popup stack](#the-popup-stack), and ending with the [nearest open ancestral popup](#nearest-open-ancestral-popup) of the `mousedown` event's `target` `Node`. This means clicking on a popup or its trigger or anchor elements will not hide that popup.

Reviewer: suggestion, maybe `mouseup` would be good here?  

I say this because [WCAG 2.5.2 Pointer Cancellation criterion](https://www.w3.org/WAI/WCAG21/Understanding/pointer-cancellation.html) could then be adhered to, allowing someone who makes an accidental down click outside of a popup (someone with motor disabilities, for instance) would have the chance to move their pointer back into the boundaries of the popup, allowing for the accidental dismissal to be aborted.

Author: This is likely worth discussion. I did a rather informal survey of native and web popups, and found that nearly all of them dismiss on mouse-down. Think it's worth an issue?

Reviewer: i think it's definitely worth at least a discussion.

there are presently a lot of platform behaviors that get a 'pass' with wcag rules because web authors can literally do nothing about them, since they're baked into the platform.  but, if given the chance to change a behavior, it at least deserves being brought up and then further discussion can dictate if there's a reason to adhere to what's been previously done vs going in a different direction.  

thanks!

Author: Filed! https://github.com/openui/open-ui/issues/529

> 2. Hitting the ESC key, or generally any ["close signal"](#close-signal). This will cause the topmost popup on [the popup stack](#the-popup-stack) to be hidden.

Reviewer: matching earlier use of the full word, rather than the abbreviated key.

```suggestion
2. Hitting the Escape key, or generally any ["close signal"](#close-signal). This will cause the topmost popup on [the popup stack](#the-popup-stack) to be hidden.
```

Author: Done

This table summarizes the interactions between a first top layer element (rows) and a second top layer element (columns), as the second element is shown:

<table>
  <tr><td></td><td></td><td colspan="5">Second element</td></tr>
  <tr><td></td><td></td><td>Fullscreen</td><td>Modal Dialog</td><td>Popup</td><td>Hint</td><td>Async</td></tr>
  <tr><td rowspan="5" >First Element</td><td>Fullscreen</td><td>Hide</td><td>Leave</td><td>Leave</td><td>Leave</td><td>Leave</td></tr>
  <tr><td>Modal Dialog</td><td>Hide*</td><td>Leave</td><td>Leave</td><td>Leave</td><td>Leave</td></tr>
  <tr><td>Popup</td><td>Hide</td><td>Hide</td><td>Hide</td><td>Leave</td><td>Leave</td></tr>
  <tr><td>Hint</td><td>Hide</td><td>Hide</td><td>Hide</td><td>Hide</td><td>Leave</td></tr>
  <tr><td>Async</td><td>Hide</td><td>Hide</td><td>Leave</td><td>Leave</td><td>Leave</td></tr>
</table>

*Not current behavior

In the table, "hide" means that when the second element is shown (enters the top layer), the first element is removed from the top layer. In contrast, "leave" means both elements will remain in the top layer together.

### Detailed description of interactions among popup types

> This section describes, in detail, the interactions between the three popup types:

Reviewer: ```suggestion
This section details the interactions between the three popup types:
```

Author: Nice

2. If a `popup=popup` is shown, it should hide **any** open `popup=hint`s, including if the `popup=hint` is an ancestral popup of the `popup=popup`. (**"You can't nest a popup inside a `popup=hint`".**)
3. If you: **a)** show a `popup=popup` (call it D), then **b)** show an **ancestral** `popup=hint` of D (call it T) , then **c)** hide D, the `popup=hint` T should be hidden. (**"A `popup=hint` can be nested inside a popup."**)
4. If you: **a)** show a `popup=popup` (call it D), then **b)** show an **non-ancestral** `popup=hint` (call it T) , then **c)** hide D, the `popup=hint` T should be left showing. (**"Non-nested `popup=hint`s can stay open when unrelated popups are hidden."**)
5. The `defaultopen` attribute should have no effect on `popup=hint`s. I.e. this attribute cannot be used to cause a `popup=hint` to be shown upon page load.
6. The `defaultopen` attribute can be used on as many `popup=async`s as desired, and all of them will be shown upon page load.
7. Only the first `popup=popup` (in DOM order) containing the `defaultopen` attribute will be shown upon page load. (This is per-explainer, and included here for completeness.)

## Accessibility / Semantics

> Since the `popup` content attribute can be applied to any element, and this only impacts the element's presentation (top layer vs not top layer), this addition does not have any direct semantic or accessibility impact. The element with the `popup` attribute will keep its existing semantics and AOM representation. For example, `<img popup=popup>` will continue to have `role=img`, but will be able to be displayed on top of other content. Similarly, ARIA can be used to modify accessibility mappings in [the normal way](https://w3c.github.io/html-aria/), for example `<img popup=popup role=presentation>`.

Reviewer: Swapping around the examples a bit because while I still think they should be valid, using an image as a popup could definitely have its use cases (i'm immediately thinking of "here's where your credit card code is"), but not the common use case I'd imagine people would go for.  

```suggestion
Since the `popup` content attribute can be applied to any element, and this only impacts the element's presentation (top layer vs not top layer), this addition does not have any direct semantic or accessibility impact. The element with the `popup` attribute will keep its existing semantics and AOM representation. For example, `<article popup=popup>...</article>` will continue to be exposed as an implicit `role=article`, but will be able to be displayed on top of other content. Similarly, ARIA can be used to modify accessibility mappings in [the normal way](https://w3c.github.io/html-aria/), for example `<div popup=popup role=note>...</div>`.
```

Author: Awesome, thanks. I struggled with a good example here - this is better.

As mentioned in the [Declarative Triggers](#declarative-triggers) section, accessibility mappings will be automatically configured to associate the popup with its trigger element, as needed.

## Disallowed elements

> While the popup API can be used on most elements, there are some limitations. For example, calling `showPopup()` on a `<dialog>` element will result in an exception being thrown, as will calling it on an [active fullscreen element](https://developer.mozilla.org/en-US/docs/Web/API/Element/requestFullscreen). All other element types are valid.

Reviewer: I do not think a dialog should be completely barred from being a popup (unless i'm misunderstanding this line?)  it's also in opposition with the example that follows in the next section, re: `<my-date-picker role=dialog id=datepicker popup=popup>`

can expand on this if need be... but maybe this should just be clarified to mean a 'modal' dialog would throw an exception?  (even though there are tons of instances of custom modal dialogs where people implement a click-outside-to-close-light-dismiss... but, i have too many other hills i've decided to die on...)

Author: I've gone back and forth on this. But I do see your point here that `role=dialog` semantics might be entirely appropriate on something that has popup behaviors (like click-outside light dismiss).

I've added explicitly "modal" dialog here.

# Additional Considerations

## Exceeding the Frame Bounds

> Allowing a popup/top-layer element to exceed the bounds of its containing frame poses a serious security risk: such an element could spoof browser UI or containing-page content. While the [original `<popup>` proposal](https://open-ui.org/components/popup.research.explainer#goals) did not discuss this issue, the [`<selectmenu>` proposal](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/ControlUICustomization/explainer.md#security) does have a specific section at least mentioning this issue. Some top-layer APIs (e.g. the fullscreen API) make it possible for an element to exceed the frame bounds in some cases, great care must be taken in these cases to ensure user safety. Given the complete flexibility offered by the Popup API (any element, arbitrary content, etc.), there would be no way to ensure the safety of this feature if it were allowed to exceed frame bounds.

Reviewer 2: Unless I've made a silly mistake, won't the following: 

`While the [original `<popup>` proposal](https://open-ui.org/components/popup.research.explainer#goals)...`

just link to the current page's goals heading? 

Is there a different document that this link should point to, or is the wording just a little ambiguous?

I'd figure that the _original <popup> proposal_ would be an older spec (maybe the abandoned one?)

Author: Great catch! That's from before I moved the explainer to here.
